### PR TITLE
Formatting postcodes

### PR DESCRIPTION
**Pull request by ali-i3**
## Pull Request submission

Fixing the issues we had with the postcode validation.
1) Fixed instances of correct postcodes being marked as incorrect
2) Output is now 2 csv files with reference and instance as lookup columns:
- A csv with incorrect pattern postcodes
- A csv with postcodes not in list
3) The pipeline doesn't fail if the postcodes are wrong, it instead outputs a file of incorrect postcodes. (Can be re-evaluated)

*Insert any instructions to help the reviewer, e.g. "install new requirements from `environment.yml`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/developer_config.yaml [hunk 1]
```diff
@@ -31,6 +32,7 @@ network_paths:
   logs_foldername: "logs/run_logs"
   snapshot_path: "R:/BERD Results System Development 2023/DAP_emulation/survey_return_data/snapshot-202212-002-83b5bacd-7c99-45cf-b989-d43d762dd054.json"
   postcode_masterlist: "R:/BERD Results System Development 2023/DAP_emulation/ONS_Postcode_Reference/ONSPD_NOV_2022_UK.csv"
+  postcode_path: "R:/BERD Results System Development 2023/DAP_emulation/staging/postcode_validation"
```
Comment (AnneONS) [inline src/developer_config.yaml]: please could this go in /staging/staging_qa/postcode_validation ?

File: src/developer_config.yaml [hunk 2]
```diff
@@ -23,6 +23,7 @@ hdfs_paths:
   logs_foldername: "resdev/logs/run_logs"
   snapshot_path: "/ons/rdbe_dev/snapshot-202012-002-fba5c4ba-fb8c-4a62-87bb-66c725eea5fd.json"
   postcode_masterlist: "/ons/rdbe_dev/mappers/2023/ONSPD_NOV_2022_UK.csv"
+  postcode_path: "/ons/rdbe_dev/inter_outputs/postcode_validation"
```
Comment (AnneONS) [inline src/developer_config.yaml]: please could you change this to /rdbe_dev/staging/staging_qa/postcode_validation ?

File: src/staging/validation.py [hunk 3]
```diff
@@ -50,9 +49,7 @@ def get_masterlist(postcode_masterlist) -> pd.Series:
```
Comment (smithv1) [inline src/staging/validation.py]: Line 21: incomplete sentence
Could add to doc string : "masterlist contains - all post codes in the UK" if that's correct
Line 39: is the doc string correct? Maybe move get_masterlist to above check_pcs_real so they can be read together more easily

File: src/staging/validation.py [hunk 4]
```diff
@@ -50,9 +49,7 @@ def get_masterlist(postcode_masterlist) -> pd.Series:
 
 @time_logger_wrap
 @exception_wrap
-def validate_post_col(
-    df: pd.DataFrame, postcode_masterlist: str
-) -> bool:
+def validate_post_col(df: pd.DataFrame, postcode_masterlist: str) -> bool:
     """This function checks if all postcodes in the specified DataFrame column
         are valid UK postcodes. It uses the `validate_postcode` function to
```
Comment (smithv1) [inline src/staging/validation.py]: update docstring validate_postcode to validate_postcode_pattern

File: src/staging/validation.py [hunk 5]
```diff
@@ -50,9 +49,7 @@ def get_masterlist(postcode_masterlist) -> pd.Series:
 
 @time_logger_wrap
 @exception_wrap
-def validate_post_col(
-    df: pd.DataFrame, postcode_masterlist: str
-) -> bool:
+def validate_post_col(df: pd.DataFrame, postcode_masterlist: str) -> bool:
```
Comment (smithv1) [inline src/staging/validation.py]: add postcode_masterlist arg

Comment (smithv1) [reply src/staging/validation.py]: Add returns to doc string - invalid_df, unreal_df

Comment (smithv1) [reply src/staging/validation.py]: Could add comment to make distiction between invalid and unreal - what distinguishes each

File: src/staging/validation.py [hunk 6]
```diff
@@ -66,50 +63,104 @@ def validate_post_col(
...
+    return invalid_df, unreal_df
+
+
+def insert_space(postcode, position):
+    return postcode[0:position] + " " + postcode[position:]
 
 
 def check_pcs_real(df: pd.DataFrame, postcode_masterlist: str):
-    """Checks if the postcodes are real against a masterlist of actual postcodes"""
+    """Checks if the postcodes are real against a masterlist of actual postcodes.
```
Comment (smithv1) [inline src/staging/validation.py:134]: Update format of doc string with args etc

File: src/staging/validation.py [hunk 7]
```diff
@@ -66,50 +63,104 @@ def validate_post_col(
...
+    All postcodes less than 7 characters have added whitespaces in the middle.
+
+    This formatting is applied to a copy dataframe so the original is unchanged.
+
+    The final output are the postcodes from the original dataframe
+    """
     if config["global"]["postcode_csv_check"]:
         master_series = get_masterlist(postcode_masterlist)
 
+        copydf = df.copy()
```
Comment (smithv1) [inline src/staging/validation.py]: maybe do .copy() at start and call the df something like check_real_df

File: src/staging/validation.py [hunk 8]
```diff
@@ -66,50 +63,104 @@ def validate_post_col(
...
+            lambda x: insert_space(x, 3)
+        )
+
         # Check if postcode are real
         unreal_postcodes = df.loc[
-            ~df["referencepostcode"].isin(master_series), "referencepostcode"
+            ~copydf["referencepostcode"].isin(master_series), "referencepostcode"
         ]
+
     else:
```
Comment (smithv1) [inline src/staging/validation.py:185]: is this for making an empty df?

File: tests/test_staging/test_validation.py [hunk 9]
```diff
@@ -24,7 +24,11 @@
```
Comment (smithv1) [inline tests/test_staging/test_validation.py:19]: Line 21: can be deleted

File: tests/test_staging/test_validation.py [hunk 10]
```diff
@@ -24,7 +24,11 @@
...
 def test_data_dict():
-    return {"referencepostcode": ["NP10 8XG", "SW1P 4DF", "HIJ 789", "KL1M 2NO"]}
+    return {
+        "reference": [1, 2, 3, 4],
+        "instance": [0, 0, 0, 0],
+        "referencepostcode": ["NP10 8XG", "SW1P 4DF", "HIJ 789", "KL1M 2NO"],
+    }
 
 
 def write_dict_to_toml(tmp_path, test_data_dict):
```
Comment (smithv1) [inline tests/test_staging/test_validation.py]: is write_dict_to_toml used, if not can be deleted?

File: tests/test_staging/test_validation.py [hunk 11]
```diff
@@ -73,47 +76,33 @@ def test_validate_post_col(test_data_df, monkeypatch, caplog):
...
     # Valid AND real postcodes
-    df_valid = pd.DataFrame({"referencepostcode": ["NP10 8XG", "PO15 5RR", "SW1P 4DF"]})
+    df_valid = pd.DataFrame(
+        {
+            "reference": [1, 2, 3],
+            "instance": [0, 0, 0],
+            "referencepostcode": ["NP10 8XG", "PO15 5RR", "SW1P 4DF"],
+        }
+    )
     assert validate_post_col(df_valid, fake_path)
```
Comment (smithv1) [inline tests/test_staging/test_validation.py]: assert on dataframes - not sure what this does

File: tests/test_staging/test_validation.py [hunk 12]
```diff
@@ -73,47 +76,33 @@ def test_validate_post_col(test_data_df, monkeypatch, caplog):
...
+    )
+    validate_post_col(df_invalid, fake_path)
+    assert "Invalid postcodes found: ['EFG 456', 'HIJ 789']" in caplog.text
 
     # Mixed valid and invalid postcodes - as is in the test_data
-    with pytest.raises(ValueError) as error:
-        validate_post_col(test_data_df, fake_path)
+
+    validate_post_col(test_data_df, fake_path)
     if config["global"]["postcode_csv_check"]:
```
Comment (smithv1) [inline tests/test_staging/test_validation.py:99]: Line 139: alternative name for unreal_postcodes: result_df
Line 156: Checking that none of the valid post codes are in unreal_postcodes - this is potentially a repetition of test_check_pcs_real_with_invalid_postcodes

**Review by smithv1 — CHANGES_REQUESTED** [review]

**Review by smithv1 — APPROVED** [review]
Code looks good, tests pass and runs fine on main with expected outputs: none in invalid patterns and 227 in missing

**Pull request merged**